### PR TITLE
Embedded closes #2531

### PR DESCRIPTION
**Pull request by shrit**
This pull request is the results of the work done in #2531.
It allows to provide cross compilation support for mlpack allowing to cross compile mlpack for any architecture that provides toolchain.

### CODE REVIEW

File: CMake/FindArmadillo.cmake [hunk 1]
```diff
@@ -91,12 +91,13 @@ if(NOT _ARMA_USE_WRAPPER OR MSVC)
...
   endif()
   if(_ARMA_USE_BLAS)
-    if(ARMADILLO_FIND_QUIETLY OR NOT ARMADILLO_FIND_REQUIRED)
-      find_package(BLAS QUIET)
+    if(NOT BLAS_FOUND)
+      if(ARMADILLO_FIND_QUIETLY OR NOT ARMADILLO_FIND_REQUIRED)
+        find_package(BLAS QUIET)
+      else()
+        find_package(BLAS REQUIRED)
+      endif()
```
Comment (shrit) [inline CMake/FindArmadillo.cmake:99]: @rcurtin, I have found the issue that stopping it from running on CI. if we compiled `OpenBLAS` manually, we need to `set(BLAS_FOUND ON)`.
The issue is related to `find_package(BLAS)` here, since it is going to search for BLAS inside the `sysroot` and will never find it (since it is in deps), then it will set `(BLAS_FOUND OFF)` stopping it from being added to `ARMADILLO_LIBRARIES`
The solution I found is just to check whether `BLAS_FOUND` is already `ON`. In this case, there is no need to run `find_package(BLAS)`

Comment (rcurtin) [reply CMake/FindArmadillo.cmake:99]: This doesn't seem right.  What is the version of CMake on CI?  Do you need to set `BLAS_LIBRARIES` instead of `BLAS_openblas_LIBRARIES`?  It shouldn't be necessary to make these changes, since they intertwine functionality of the cross-compilation CMake configuration and this.

File: CMake/ConfigureCrossCompile.cmake [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  find_package(BLAS)
+  if (NOT BLAS_FOUND OR (NOT BLAS_LIBRARIES))
+    get_deps(https://github.com/xianyi/OpenBLAS/releases/download/v${version}/OpenBLAS-${version}.tar.gz OpenBLAS OpenBLAS-${version}.tar.gz)
+    if (NOT MSVC)
+      if (NOT EXISTS "${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version}/libopenblas.a")
+        execute_process(COMMAND make TARGET=${OPENBLAS_TARGET} BINARY=${OPENBLAS_BINARY} HOSTCC=gcc CC=${CMAKE_C_COMPILER} FC=${CMAKE_FORTRAN_COMPILER} NO_SHARED=1
+                        WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version})
+      endif()
+      file(GLOB OPENBLAS_LIBRARIES "${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version}/libopenblas.a")
+      set(BLAS_LIBRARIES ${OPENBLAS_LIBRARIES})
```
Comment (rcurtin) [inline CMake/ConfigureCrossCompile.cmake:25]: I think you may also (or instead) need to set `BLAS_openblas_LIBRARIES` here.

**Review by rcurtin — COMMENTED** [review]
Looks good to me; if we can clean up the Armadillo/BLAS searching then I think this is good for merge. :+1:  (Sorry it took a couple days to get to a review for this.)

File: CMake/ConfigureCrossCompile.cmake [hunk 3]
```diff
@@ -0,0 +1,33 @@
+if (CMAKE_CROSSCOMPILING)
```
[the remark this answers is not in the document]
Comment (shrit) [reply CMake/ConfigureCrossCompile.cmake:1]: Of course, I will add some comments here :+1:

File: CMake/ConfigureCrossCompile.cmake (hunk 2 above)
Comment (shrit) [reply CMake/ConfigureCrossCompile.cmake:25]: As you wish.

File: CMake/FindArmadillo.cmake (hunk 1 above)
Comment (shrit) [reply CMake/FindArmadillo.cmake:99]: The version of CMake on CI is 3.13.4

This is the only solution I have found, but I am open to find a better solution in order to reduce the modifications to cmake files :+1:.

After hours of digging, I have discovered that it is not related to the version of cmake, even I was able to reproduce the same issue on my machine. It will work only using `BLAS_LIBRARIES` when I run cmake twice (first time failing BLAS not found) and keep the cache and then run `find_package(BLAS)`.

I do think you will face the same issue if you remove OpenBLAS from the sysroot, let me know if this is not the case.

If my understanding is right `find_package(BLAS)` will start by resetting the value of `BLAS_LIBRARIES` https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/FindBLAS.cmake#L283

Then it will look inside the sysroot (when cross comiling), if it is not there it will not find it.

I have already set `BLAS_LIBRARIES` and `BLAS_openblas_LIBRARIES` which gave the same result.

I happy to remove these changes and run it on CI a second time if you like in order to repoduce the issue.

Let me know what do you think :+1:

File: CMake/ConfigureCrossCompile.cmake [hunk 4]
```diff
@@ -0,0 +1,42 @@
+# This file adds the necessary configurations to cross compile
+# mlpack for embedded systems. You need to set the following variables
+# from the command line: the CMAKE_SYSROOT, TOOLCHAIN_PREFIX and the
+# board type.
+# This file will compile OpenBLAS if it is downloaded and it is not
+# available on you system in order to find the BLAS library.
```
Comment (zoq) [inline CMake/ConfigureCrossCompile.cmake:6]:
```suggestion
# available on your system in order to find the BLAS library.
```

Minor grammer fix.

File: CMake/ConfigureCrossCompile.cmake [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+if (CMAKE_CROSSCOMPILING)
+  include(board/flags-config.cmake)
+  if(NOT CMAKE_SYSROOT AND (NOT TOOLCHAIN_PREFIX))
+    message(FATAL_ERROR "Neither of CMAKE_SYSROOT or TOOLCHAIN_PREFIX is set, please set both of them and try again")
+  elseif(NOT CMAKE_SYSROOT)
+    message(FATAL_ERROR "Can not proceed CMAKE_SYSROOT is not set")
+  elseif(NOT TOOLCHAIN_PREFIX)
+    message(FATAL_ERROR "Cant not proceed TOOLCHAIN_PREFIXN is not set")
```
Comment (zoq) [inline CMake/ConfigureCrossCompile.cmake:15]:
```suggestion
    message(FATAL_ERROR "Cant not proceed TOOLCHAIN_PREFIX is not set")
```

Use the correct name.

File: CMake/ConfigureCrossCompile.cmake [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  elseif(NOT OPENBLAS_TARGET)
+    message(FATAL_ERROR "Cant not proceed, board name is not set, please refer to documentation")
```
Comment (zoq) [inline CMake/ConfigureCrossCompile.cmake:17]: Might be a good idea to include a link to the documentation here, I guess even linking to https://mlpack.org/docs.html would be helpful?

File: CMake/ConfigureCrossCompile.cmake [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    get_deps(https://github.com/xianyi/OpenBLAS/releases/download/v${version}/OpenBLAS-${version}.tar.gz OpenBLAS OpenBLAS-${version}.tar.gz)
+    if (NOT MSVC)
+      if (NOT EXISTS "${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version}/libopenblas.a")
+        execute_process(COMMAND make TARGET=${OPENBLAS_TARGET} BINARY=${OPENBLAS_BINARY} HOSTCC=gcc CC=${CMAKE_C_COMPILER} FC=${CMAKE_FORTRAN_COMPILER} NO_SHARED=1
+                        WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version})
+      endif()
+      file(GLOB OPENBLAS_LIBRARIES "${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version}/libopenblas.a")
+      set(BLAS_LIBRARIES ${OPENBLAS_LIBRARIES})
+      set(BLAS_openblas_LIBRARIES ${OPENBLAS_LIBRARIES})
+      message(STATUS "SHOW BLAS libraries: ${BLAS_LIBRARIES}")
```
Comment (zoq) [inline CMake/ConfigureCrossCompile.cmake:34]: I would probably remove the `SHOW` here, we are printing it so showing is somewhat implicit?

File: CMake/FindArmadillo.cmake [hunk 8]
```diff
@@ -91,11 +91,12 @@ if(NOT _ARMA_USE_WRAPPER OR MSVC)
     endif()
   endif()
   if(_ARMA_USE_BLAS)
-    if(ARMADILLO_FIND_QUIETLY OR NOT ARMADILLO_FIND_REQUIRED)
-      find_package(BLAS QUIET)
-    else()
-      find_package(BLAS REQUIRED)
-    endif()
+#    if(NOT BLAS_FOUND)
```
Comment (zoq) [inline CMake/FindArmadillo.cmake:94]: Looks like that line can be removed?

File: CMake/ConfigureCrossCompile.cmake [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    message(FATAL_ERROR "Cant not proceed TOOLCHAIN_PREFIXN is not set")
+  elseif(NOT OPENBLAS_TARGET)
+    message(FATAL_ERROR "Cant not proceed, board name is not set, please refer to documentation")
+  endif()
+endif()
+
+macro(search_openblas version)
+  set(BLA_STATIC ON)
+  find_package(BLAS)
+  if (NOT BLAS_FOUND OR (NOT BLAS_LIBRARIES))
```
Comment (zoq) [inline CMake/ConfigureCrossCompile.cmake:24]:
```suggestion
  if(NOT BLAS_FOUND OR (NOT BLAS_LIBRARIES))
```

Let's use the same style across the file.

File: board/crosscompile-toolchain.cmake [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+set(CMAKE_SYSROOT "" CACHE STRING "CMAKE_SYSROOT")
+set(TOOLCHAIN_PREFIX "" CACHE STRING "TOOLCHAIN_PREFIX")
+
+## In some distribution, a dynamic link for aarch64-linux-gnu-gcc may not be
+## found or created, instead it might be labeled with the version at the end
+## For instance: aarch64-linux-gnu-gcc-5
+## Therefore, if dynamic link exists, you do not have to specify the version
+set(VERSION_NUMBER "" CACHE STRING "Enter the version number of the compiler")
+
+# Without that flag CMake is not able to pass test compilation check
```
Comment (zoq) [inline board/crosscompile-toolchain.cmake:18]:
```suggestion
# Without that flag CMake is not able to pass test compilation check.
```

File: board/crosscompile-toolchain.cmake [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+## command line.
+
+set(CMAKE_SYSTEM_NAME Linux)
+set(CMAKE_SYSROOT "" CACHE STRING "CMAKE_SYSROOT")
+set(TOOLCHAIN_PREFIX "" CACHE STRING "TOOLCHAIN_PREFIX")
+
+## In some distribution, a dynamic link for aarch64-linux-gnu-gcc may not be
+## found or created, instead it might be labeled with the version at the end
+## For instance: aarch64-linux-gnu-gcc-5
+## Therefore, if dynamic link exists, you do not have to specify the version
```
Comment (zoq) [inline board/crosscompile-toolchain.cmake:15]:
```suggestion
## Therefore, if dynamic link exists, you do not have to specify the version.
```

File: board/crosscompile-toolchain.cmake [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+set(CMAKE_LINKER ${TOOLCHAIN_PREFIX}ld${VERSION_NUMBER})
+set(CMAKE_C_ARCHIVE_CREATE "<CMAKE_AR> qcs <TARGET> <LINK_FLAGS> <OBJECTS>")
+set(CMAKE_C_ARCHIVE_FINISH  true)
+set(CMAKE_FORTRAN_COMPILER ${TOOLCHAIN_PREFIX}gfortran)
+set(CMAKE_ASM_COMPILER ${CMAKE_C_COMPILER})
+set(CMAKE_OBJCOPY ${TOOLCHAIN_PREFIX}objcopy${VERSION_NUMBER} CACHE INTERNAL "objcopy tool")
+set(CMAKE_SIZE_UTIL ${TOOLCHAIN_PREFIX}size${VERSION_NUMBER} CACHE INTERNAL "size tool")
+
+## Here are the standard ROOT_PATH if you are using the standard toolchain
+## if you are using a different toolchain you have to specify that too
```
Comment (zoq) [inline board/crosscompile-toolchain.cmake:33]:
```suggestion
## if you are using a different toolchain you have to specify that too.
```

File: CMake/FindArmadillo.cmake (hunk 8 above)
Comment (shrit) [reply CMake/FindArmadillo.cmake:94]: Just commented out, to see if it can work without.

File: CMake/ConfigureCrossCompile.cmake (hunk 7 above)
Comment (shrit) [reply CMake/ConfigureCrossCompile.cmake:34]: Yeah, it is still not working, I will remove these debugging symbols later, I just have added them

File: board/flags-config.cmake [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+option(RPI0 "Optimize compiler flags for Raspberry PI 0." OFF)
+option(RPI1 "Optimize compiler flags for Raspberry PI 1." OFF)
+option(RPI2 "Optimize compiler flags for Raspberry PI 2." OFF)
+option(RPI3 "Optimize compiler flags for Raspberry PI 3." OFF)
+option(RPI4 "Optimize compiler flags for Raspberry PI 4." OFF)
+option(BV "Optimize compiler flags for Beagleboard V." OFF)
+option(JETSONAGX "Optimize compiler flags for Nvidia Jetson AGX Xavier." OFF)
+option(KATAMI "Optimize compiler flags for Pentium 3 Katami processors." OFF)
+option(COPPERMINE "Optimize compiler flags for Pentium 3 Coppermine processors." OFF)
+option(NORTHWOOD "Optimize compiler flags for Pentium 4 Northwood processors." OFF)
```
Comment (rcurtin) [inline board/flags-config.cmake:27]: Personally I think this is better as a single string option like `BOARD_NAME`, like you had it before.  The reason I say that is because with ten different `option()`s, the user could set all of them to true, but they have no way of knowing from the documentation what the precedence order is.  Since it doesn't make sense to set multiple boards (and since you are using an `if`/`elseif` structure below), I think it would be a lot cleaner to just use `BOARD_NAME` like before. :+1:

File: CMake/ConfigureCrossCompile.cmake (hunk 6 above)
Comment (shrit) [reply CMake/ConfigureCrossCompile.cmake:17]: That is true, I will add the link :+1: 
But we really need to update documentation later, since there is nothing there related to this one :+1:

File: board/flags-config.cmake (hunk 13 above)
Comment (shrit) [reply board/flags-config.cmake:27]: Ok, I will change that back, I just hope that the user does not misspell the board name

File: CMake/FindArmadillo.cmake (hunk 8 above)
Comment (rcurtin) [reply CMake/FindArmadillo.cmake:94]: This should be able to work without any changes to `FindArmadillo.cmake`.  I'm sorry to dig in my feet so hard about this issue; basically, I think something is poorly designed if it intertwines two components that don't need to be intertwined.  In this case, the change being made to `FindArmadillo.cmake` is specific to the cross-compilation toolchain... but any CMake `find_package` script is supposed to be generic enough to be used in any situation.  So, if we have code in this generic `find_package` script specific to our cross-compilation setup, we are setting ourselves up for a very difficult debugging time later, since we have introduced some code specific to some exceptional situation that only happens in our cross-compilation setup.  I can virtually guarantee that at some point in the future, that `if (NOT BLAS_FOUND)` condition will get activated in someone else's setup when they are not using cross-compilation, and that will cause all kind of headaches to debug.

Sorry to write so much and belabor the point; I hope my motivations make sense.  I am debugging the issue right now to try and find a solution. :+1:

Comment (shrit) [reply CMake/FindArmadillo.cmake:94]: I totally agree, I understand the issue.
I left you a comment above with a link to `Findblas.cmake` to show where the issue is coming from.

File: CMake/ConfigureCrossCompile.cmake (hunk 2 above)
Comment (rcurtin) [reply CMake/ConfigureCrossCompile.cmake:32]: I was remembering the name of the variable from memory incorrectly.  It is actually `BLAS_openblas_LIBRARY`, and when you set this, even when OpenBLAS is not installed in the buildroot environment, configuration and then build/linking succeeds.

I spent some time debugging and I think I have a conclusion but I have to go cook dinner now...  Here's what I found:

 * I am not sure what the case is with older CMake, but at least with newer CMake I don't need to set BLAS_LIBRARIES.  I think you should try removing it for older CMake and see what happens with the older version of FindBLAS.cmake.  (I forget at what CMake version this changed, but I think we wrote down the version number in a chat somewhere or something.)

 * I found that there is some strange interplay between `FindLAPACK.cmake` and `FindBLAS.cmake`.  It seems in many cases, `FindLAPACK.cmake` will call out to `FindBLAS.cmake`, and if a joint implementation of LAPACK/BLAS is found (which is the case for OpenBLAS), that will be used.  But in order to make the scripts find it, I had to set `LAPACK_openblas_LIBRARY` just like we set `BLAS_openblas_LIBRARY`.  I believe if you set those two variables (as well as `BLA_VENDOR` and `BLA_STATIC`), configuration should succeed on newer CMake.  (Older CMake might require `BLAS_LIBRARIES` and `LAPACK_LIBRARIES`... I'd try without those first, and if that fails, try adding them in?)

 * Once those changes above are done, you should be able to revert `FindArmadillo.cmake` to its original state in the master branch.

I think that's it... let me know what you find out if you try it.  For now it's dinner time. :)

Comment (shrit) [reply CMake/ConfigureCrossCompile.cmake:32]: Each test done after cleaning caching, here is my matrix build:
Localy
set only `BLAS_LIBRARIES` does not work
set only `BLAS_openblas_LIBRARY` works
set `BLAS_openblas_LIBRARY` and `BLAS_LIBRARIES` works (thanks to the `BLAS_openblas_LIBRARY`)
set only `LAPACK_openblas_LIBRARY` does not work
set `BLAS_openblas_LIBRARY` and `LAPACK_openblas_LIBRARY` works (thanks to the `BLAS_openblas_LIBRARY`)
set `BLAS_LIBRARIES` and `LAPACK_LIBRARIES` does not work
on CI:
set only `BLAS_LIBRARIES` does not work
set only `BLAS_openblas_LIBRARY` does not work
set `BLAS_openblas_LIBRARY` and `BLAS_LIBRARIES` does not work
set only `LAPACK_openblas_LIBRARY` does not work
set `BLAS_openblas_LIBRARY` and `LAPACK_openblas_LIBRARY`  does not work
set `BLAS_LIBRARIES` and `LAPACK_LIBRARIES` does not work

Maybe it is the time for a new version of CMake on CI from the current one 3.13.4?

Comment (rcurtin) [reply CMake/ConfigureCrossCompile.cmake:32]: CMake 3.13.4 (and newer versions that still have the old FindBLAS/FindLAPACK) will still be present in the wild, so we can't upgrade.  We will have to find a solution.  I guess I will dig further tomorrow.

File: board/flags-config.cmake (hunk 13 above)
Comment (rcurtin) [reply board/flags-config.cmake]: Well, you can catch that situation pretty easily with the `if`/`elseif` block; just add an `else` at the end if the board isn't recognized. :)

File: CMake/ConfigureCrossCompile.cmake (hunk 2 above)
Comment (rcurtin) [reply CMake/ConfigureCrossCompile.cmake:32]: Using CMake version 3.10.2 seems to work by setting `BLAS_openblas_LIBRARY` and `LAPACK_openblas_LIBRARY`.

Comment (rcurtin) [reply CMake/ConfigureCrossCompile.cmake:32]: Indeed I was even able to use `BLAS_openblas_LIBRARY` and `LAPACK_openblas_LIBRARY` successfully on CMake 3.13.4.  I simply uncommented the lines setting those variables in `ConfigureCrossCompile.cmake`:

```
root@0a5042e2b194:/src/mlpack/build-arm64# git diff
diff --git a/CMake/ConfigureCrossCompile.cmake b/CMake/ConfigureCrossCompile.cmake
index 3078d400d..d0f1fa1b0 100644
--- a/CMake/ConfigureCrossCompile.cmake
+++ b/CMake/ConfigureCrossCompile.cmake
@@ -31,8 +31,8 @@ macro(search_openblas version)
       file(GLOB OPENBLAS_LIBRARIES "${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version}/libopenblas.a")
       set(BLAS_LIBRARIES ${OPENBLAS_LIBRARIES})
       set(LAPACK_LIBRARIES ${OPENBLAS_LIBRARIES})
-      # set(BLAS_openblas_LIBRARY ${OPENBLAS_LIBRARIES})
-      # set(LAPACK_openblas_LIBRARY ${OPENBLAS_LIBRARIES})  
+      set(BLAS_openblas_LIBRARY ${OPENBLAS_LIBRARIES})
+      set(LAPACK_openblas_LIBRARY ${OPENBLAS_LIBRARIES})  
       message(STATUS "SHOW BLAS libraries: ${BLAS_LIBRARIES}")
       set(BLA_VENDOR OpenBLAS)
       set(BLAS_FOUND ON)
```

I suspect that `BLAS_LIBRARIES` and `LAPACK_LIBRARIES` don't need to be set.  Can you double-check to see if you see the same results on your setup?

File: CMake/ConfigureCrossCompile.cmake (hunk 6 above)
Comment (rcurtin) [reply CMake/ConfigureCrossCompile.cmake:17]: Actually, I am not sure that not having `OPENBLAS_TARGET` set is really a problem.  That would only be a problem if we need to autodownload and compiled OpenBLAS ourselves.  So perhaps you should move this check there.  Also, I would suggest using an error message that says that `OPENBLAS_TARGET` is not set, and to either set that or `BOARD_NAME` (which will automatically set `OPENBLAS_TARGET`).

File: CMakeLists.txt [hunk 14]
```diff
@@ -265,6 +269,10 @@ endif()
 #   STB_IMAGE_INCLUDE_DIR - include directory for STB image library
 #   MATHJAX_ROOT - root of MathJax installation
 
+if (CMAKE_CROSSCOMPILING)
```
Comment (rcurtin) [inline CMakeLists.txt:272]: It could be nice to add a comment here, even though it may seem obvious in the context of this PR.  Later maintainers (maybe even us if we forget about this months from now :smile:) might not know why this is needed.

File: board/crosscompile-toolchain.cmake [hunk 15]
```diff
@@ -0,0 +1,41 @@
+## This file handles cross-compilation configurations for aarch64,
+## known as arm64. The objective of this file is to find and assign
+## cross-compiler and the entire toolchain.
+## It works best with buildroot toolchain, when using it the user
```
Comment (rcurtin) [inline board/crosscompile-toolchain.cmake:4]: It could be useful to add a link to the buildroot website here.  (I didn't know what buildroot was before we started on these PRs.)

File: board/flags-config.cmake [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -findirect-inlining -fno-common") 
+#-flto -fuse-ld=gold # There is an issue with gold link when compiling on 
+# Ubuntu 16. At that point gcc linker did not integrate the flto support 
+# inside and it was a separate plugin that need to be added. Therefore, 
+# this can be added when mlpack Azure CI moves toward Ubuntu 20.
+
+set(BOARD_NAME "" CACHE STRING "Specify Board name to optimize for.")
+
+# Set specific platforms CMAKE CXX flags.
+if(BOARD_NAME MATCHES "RPI0" OR BOARD_NAME MATCHES "RPI1")
```
Comment (rcurtin) [inline board/flags-config.cmake:21]: It might be nice to use `string(TOUPPER ...)` on `BOARD_NAME` before these comparisons.  (I'd suggest doing that into a temporary variable, so that we don't change whatever the user specified as `BOARD_NAME`.)  This way, the user can specify `RPI1` or `rpi1` or even `rPi1` and there will be no problem. :+1:

File: board/crosscompile-toolchain.cmake [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+## needs to set the: TOOLCHAIN_PREFIX and CMAKE_SYSROOT from the
+## command line.
+
+set(CMAKE_SYSTEM_NAME Linux)
+set(CMAKE_SYSROOT "" CACHE STRING "CMAKE_SYSROOT")
```
Comment (rcurtin) [inline board/crosscompile-toolchain.cmake:9]: I don't think this one needs to be set as a CACHE variable.

File: board/crosscompile-toolchain.cmake [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+set(TOOLCHAIN_PREFIX "" CACHE STRING "TOOLCHAIN_PREFIX")
```
Comment (rcurtin) [inline board/crosscompile-toolchain.cmake:10]: Do you think you can clean up the CACHE documentation strings?  They should be, e.g., a short sentence describing what the variable represents.  (A user using the CMake GUI may see these strings, for instance.)  As an example, for this one, I might suggest `Toolchain-specific prefix for compiler and other compiler-related tools`.  (Maybe that could be improved; it's just a suggestion. :))

File: CMake/ConfigureCrossCompile.cmake [hunk 19]
```diff
@@ -0,0 +1,42 @@
+# This file adds the necessary configurations to cross compile
+# mlpack for embedded systems. You need to set the following variables
+# from the command line: the CMAKE_SYSROOT, TOOLCHAIN_PREFIX and the
+# board type.
```
Comment (rcurtin) [inline CMake/ConfigureCrossCompile.cmake:4]:
```suggestion
# from the command line: CMAKE_SYSROOT and TOOLCHAIN_PREFIX.
```
Technically BOARD_NAME does not need to be set always.

File: CMake/ConfigureCrossCompile.cmake [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+# This file will compile OpenBLAS if it is downloaded and it is not
+# available on your system in order to find the BLAS library.
```
Comment (rcurtin) [inline CMake/ConfigureCrossCompile.cmake:6]:
```suggestion
# available on your system in order to find the BLAS library.  If OpenBLAS will
# be compiled, the OPENBLAS_TARGET variable must be set.  This can be done
# by, e.g., setting BOARD_NAME (which will set OPENBLAS_TARGET in
# `board/flags-config.cmake`).
```

File: CMake/FindArmadillo.cmake [hunk 21]
```diff
@@ -91,11 +91,12 @@ if(NOT _ARMA_USE_WRAPPER OR MSVC)
...
+      if(ARMADILLO_FIND_QUIETLY OR NOT ARMADILLO_FIND_REQUIRED)
+        find_package(BLAS QUIET)
+      else()
+        find_package(BLAS REQUIRED)
+      endif()
```
Comment (rcurtin) [inline CMake/FindArmadillo.cmake:99]: I think you can revert the changes in this file altogether now. :+1:

File: board/crosscompile-toolchain.cmake [hunk 22]
```diff
@@ -0,0 +1,38 @@
+## This file handles cross-compilation configurations for aarch64,
+## known as arm64. The objective of this file is to find and assign
+## cross-compiler and the entire toolchain.
+## It works best with buildroot toolchain, when using it the user
+## needs to set the: TOOLCHAIN_PREFIX and CMAKE_SYSROOT from the
+## command line.
```
Comment (rcurtin) [inline board/crosscompile-toolchain.cmake:6]:
```suggestion
##
## This configuration works best with the buildroot toolchain.  When using this
## file, be sure to set the TOOLCHAIN_PREFIX and CMAKE_SYSROOT variables,
## preferably via the CMake configuration command (e.g. `-DCMAKE_SYSROOT=<...>`).
##
```

This might be a little clearer for users; let me know what you think. :+1:

File: board/crosscompile-toolchain.cmake [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+## Currently, we recommend using buildroot toolchain for
+## cross-compilation. Here is the link to download the toolchains:
+## https://toolchains.bootlin.com/
+
+set(CMAKE_SYSTEM_NAME Linux)
+set(CMAKE_SYSROOT)
+set(TOOLCHAIN_PREFIX "" CACHE STRING "Path for Toolchain for cross compiler and other compilation tools.")
+
+# Without that flag CMake is not able to pass test compilation check.
```
Comment (rcurtin) [inline board/crosscompile-toolchain.cmake:15]:
```suggestion
# Ensure that CMake tries to build static libraries when testing the compiler.
```
This might be a little more accurate---up to you if you want to change it. :+1:

File: board/flags-config.cmake [hunk 24]
```diff
@@ -0,0 +1,66 @@
+# This function provides a set of specific flags for each supported board
+# Depending on the processor type. The objective is to optimize for size.
```
Comment (rcurtin) [inline board/flags-config.cmake:2]:
```suggestion
# depending on the processor type. The objective is to optimize for size.
```
(I think the `depending` goes with the previous line; correct me if I'm wrong.)

File: board/flags-config.cmake [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+# Thus, all of the fllowing flags are chosen carefully to reduce binary
```
Comment (rcurtin) [inline board/flags-config.cmake:3]:
```suggestion
# Thus, all of the following flags are chosen carefully to reduce binary 
```

File: board/flags-config.cmake [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+elseif(BOARD MATCHES "COPPERMINE")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=pentium3")
+  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,--gc-sections")
+  set(OPENBLAS_TARGET "COPPERMINE")
+  set(OPENBLAS_BINARY "32")
+elseif(BOARD MATCHES "NORTHWOOD")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=pentium4")
+  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,--gc-sections")
+  set(OPENBLAS_TARGET "NORTHWOOD")
+  set(OPENBLAS_BINARY "32")
```
Comment (rcurtin) [inline board/flags-config.cmake:65]: I think it would be worth adding an `else()` here to ensure that the user hasn't specified an invalid board.  But, you might need to do `elseif(BOARD)` so that an error isn't issued when the user didn't specify `BOARD_NAME` at all.  I'm not sure on that one, so maybe it's worth playing with to see. :+1:

**Review by rcurtin — COMMENTED** [review]
I probably should have done a review instead of leaving individual comments, so, sorry if you got like 20 emails about it. :)  But, I think once you handle these, from my side everything seems good to go. :smile:

File: board/flags-config.cmake (hunk 25 above)
Comment (shrit) [reply board/flags-config.cmake:3]: I really need a good keyboard, a true difference when using a laptop keyboard and my dell PC

**Comment by shrit** [comment]
no worries, I have applied all the modifications.

File: board/flags-config.cmake (hunk 26 above)
Comment (shrit) [reply board/flags-config.cmake:65]: Exactly, it needs `elseif(BOARD)` :+1:

File: board/flags-config.cmake [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+elseif(BOARD)
+  message(FATAL_ERROR "Board type is not known, please choose a supported board from the list")
```
Comment (zoq) [inline board/flags-config.cmake:67]:
```suggestion
  message(FATAL_ERROR "Board type is not known, please choose a supported board from the list.")
```

**Review by zoq — COMMENTED** [review]
One last really minor remark about the style.

**Review by rcurtin — COMMENTED** [review]
Ah, oops, I forgot to hit 'comment' on this review.  Some of these might conflict with suggestions @zoq posted more recently---pick whichever you think is better if that's the case. :+1:

File: board/flags-config.cmake (hunk 27 above)
[the remark this answers is not in the document]
Comment (shrit) [reply board/flags-config.cmake:67]: We do not have documentation yet, but we can add the available list somewhere, let me know where to do that, or we can wait to add the entire list when the CI matrix is done and the documentation are written :+1:

[the remark this answers is not in the document]
Comment (rcurtin) [reply board/flags-config.cmake]: That works---want to just add a TODO here or something to mention that we need to point the user at the list of boards somehow?  Then we can remove the TODO when we get all the CI work finished.

In any case we should change the error message to use the term `BOARD_NAME` (actually it is `BOARD_NAME` not `BOARD_TYPE`, my suggestion above is wrong), so that it's clear we're talking about a CMake variable of the same name. :+1:

File: board/flags-config.cmake [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,--gc-sections")
+  set(OPENBLAS_TARGET "COPPERMINE")
+  set(OPENBLAS_BINARY "32")
+elseif(BOARD MATCHES "NORTHWOOD")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=pentium4")
+  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,--gc-sections")
+  set(OPENBLAS_TARGET "NORTHWOOD")
+  set(OPENBLAS_BINARY "32")
+elseif(BOARD)
+  message(FATAL_ERROR "Given BOARD_TYPE is not known; please choose a supported board from the list")
```
Comment (rcurtin) [inline board/flags-config.cmake:67]:
```suggestion
  message(FATAL_ERROR "Given BOARD_NAME is not known; please choose a supported board from the list")
```
Oops, the variable is called `BOARD_NAME`. :)

**Review by rcurtin — APPROVED** [review]
Awesome, everything looks good to me!  It looks like it passes the CI checks too.

File: board/flags-config.cmake (hunk 27 above)
[the remark this answers is not in the document]
Comment (shrit) [reply board/flags-config.cmake]: Of course, I will add a todo here.

**Review by zoq — APPROVED** [review]
No more comments from my side, this looks great :+1:

**Comment by rcurtin** [comment]
Awesome, glad to get this finally merged. :rocket:

**Comment by shrit** [comment]
Thanks for merging it:+1:
